### PR TITLE
used typed keys to pass the current context lang to scala (#8288)

### LINKED ISSUE #8288: [2.6.x] Http.Context.messages() makes hard to override.
https://github.com/playframework/playframework/blob/191cdf004ccc3fa4fa5d255d08662b43d7ccec20/framework/src/play/src/main/java/play/mvc/Http.java#L228

The `messages()` API has changed.

In Play 2.4, the `messages()` API is just a simple wrap-around `MessagesApi.preferred(Request)`, which I could override and provide different logic.

With Play 2.6, the `messages()` API contains its own logic and no longer relies on `MessagesApi.preferred(request)`. So it makes very hard to override. Now I can only override `MessagesApi.preferred(candidates)` with calling `Http.Context.current()`, which is bad to use global state.

I think, the `messages()` API should either revert to wrapping `MessagesApi.preferred(request)`, or have `MessagesApi.preferred(candidates)` with an extra `Request` parameter.

### PR DESCRIPTION
**Pull request by schmitch**
# Helpful things

## Fixes

Fixes #8288

## Purpose

Currently in Play 2.6.x we tried to have a change that adds the current context language from Java to the MessagesApi in Scala, however due to bin incompatible changes it got reverted here: https://github.com/playframework/playframework/commit/c7855328e14f7f99d68f8d5ce072e1bbe5f26cc6

## Background Context

This addresses it by using a TypedKey, which has the benefit of not breaking any api contract and also not introducing any kind of binary incompability.
It also saves duplicated code.

### CODE REVIEW

**Comment by mkurz** [comment]
Please wait merging this pull request, I am working towards enhancements for the Java Http.Context, which might is useful for here as well.

**Comment by schmitch** [comment]
@mkurz would your change replace the enhancement or would it be a addition? because I think this PR would be great to have at least in 2.6.x even if your PR would change stuff for 2.7.x

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala [hunk 1]
```diff
@@ -34,6 +35,8 @@ import scala.util.parsing.input._
  */
 object Messages extends MessagesImplicits {
 
+  private[play] val currentContextLang = TypedKey.apply[Lang]("contextLang")
```
Comment (marcospereira) [inline framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Not a big deal, but we are using this pattern in most places to declare TypedKeys:

https://github.com/playframework/playframework/blob/4c70b05ca98227474b612bff6fccd6ae94b73b42/framework/src/play/src/main/scala/play/api/routing/Router.scala#L107-L115

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala [hunk 2]
```diff
@@ -457,8 +460,9 @@ class DefaultMessagesApi @Inject() (
   }
 
   override def preferred(request: RequestHeader): Messages = {
+    val maybeLangFromContext = request.attrs.get(Messages.currentContextLang)
     val maybeLangFromCookie = request.cookies.get(langCookieName).flatMap(c => Lang.get(c.value))
-    val lang = langs.preferred(maybeLangFromCookie.toSeq ++ request.acceptLanguages)
+    val lang = langs.preferred(maybeLangFromContext.toSeq ++ maybeLangFromCookie.toSeq ++ request.acceptLanguages)
```
Comment (marcospereira) [inline framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: I think we need to change the order here to keep the existing behavior. Before, cookie lang had the precedence over the transient language (used to populate the context's lang).

**Comment by marcospereira** [comment]
@schmitch it is unclear to me how this is fixing #8288. Is the idea here to simplify the implementation to make it easier to override?

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala (hunk 2 above)
Comment (schmitch) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: well I actually tought that this is the correct order. But actually changing it does not break any tests, maybe @mkurz could say the correct order?

**Comment by schmitch** [comment]
first it simplifies it, second it's not breaking backwards incompability and it makes it easier to override since you just need to create a new injection point for the `MessagesApi`.

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala (hunk 1 above)
Comment (schmitch) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: great suggestion, the only question would be to make it public. Because basically it's only useful for the java people.

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala (hunk 2 above)
Comment (marcospereira) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: > But actually changing it does not break any tests

Time to add some tests then? I think that makes sense to use cookie at first since this would be usually set explicitly at some point to override the request's language. It is more like "use the request lang unless I tell you to use another one" and the way to "tell" is the cookie.

WDYT?

Comment (schmitch) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Yes, there need to be some tests, I tought that there are, so I hoped it would fail if I switch it. ;)

Well what you say makes perfect sense, however the current implementation is like that:
```
LinkedList<Lang> langs = Lists.newLinkedList(request().acceptLanguages());
if (cookieLang != null) {
 langs.addFirst(cookieLang);
 }
if (lang != null) {
langs.addFirst(lang);
}
return messagesApi().preferred(langs);
```

where `lang` is the context language, so it actually is at the front of the list.
I guess the idea was that if people set a context lang that they could actually set the context lang and than use `messages()` so something like that (pseudo-code):
```
public Result langSwitcher() {
 JsonNode json = request().body().asJson();
String lang = json.get("lang").asText()
if (lang != null) {
ctx().changeLang(lang)
}
return userSettingsFormView(ctx().messages());
}
```

So basically I tought that if a user would change his language setting that the new form picks up the new lang without adding the context lang to the view.
But I guess @mkurz could probably clarify it.

Comment (mkurz) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Alright, so I just had a look at the diff before even looking at this comment and I am pretty sure that the correct order should be `val lang = langs.preferred(maybeLangFromContext.toSeq ++ maybeLangFromCookie.toSeq ++ request.acceptLanguages)` with priority from left to right, so context lang is choosen over cookie lang which is choosen over the acceptLanguages lang.
The whole point of the (transient) contextLang ist to overrule all other langs just for a *single* request/response - read that javadoc comment [here](https://github.com/playframework/playframework/blob/2.6.12/framework/src/play/src/main/java/play/mvc/Http.java#L318-L321). Another proof for this: Look at the `ctx.lang()` [method](https://github.com/playframework/playframework/blob/2.6.12/framework/src/play/src/main/java/play/mvc/Http.java#L218-L219), it will return the contextLang if it is set, not caring about a cookie or something. And that `.lang()` method is what you use when accessing the language of your request/context, in the end-user code or Play itself e.g. to localize forms [here](https://github.com/playframework/playframework/blob/2.6.12/framework/src/play-java-forms/src/main/java/play/data/Form.java#L734) and [here](https://github.com/playframework/playframework/blob/2.6.12/framework/src/play-java-forms/src/main/java/play/data/Form.java#L1054).

There should definitely be tests for to verify this order.

Comment (mkurz) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Maybe `HttpTest.java` is the place to add those tests? I added related checks two years ago there:
https://github.com/playframework/playframework/commit/a6f79c64b5b98b074279e4fc378a2ee824e06dff#diff-94832100633fa90aaf49fbe8ef37407c

Comment (mkurz) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: @schmitch So I am ok with that PR if you change and test the order of the langs.

**Comment by schmitch** [comment]
added a test and corrected the ordering like @mkurz told us

File: framework/src/play-java/src/test/java/play/mvc/HttpTest.java [hunk 3]
```diff
@@ -78,6 +78,31 @@ public void testChangeLang() {
...
+                    contextComponents);
+
+            // if no context lang is provided the language order will be cookie > accept language
+            assertThat(ctx2.messages().lang().code()).isEqualTo("fr");
+
+            // if a context lang is set the language order will be context lang > cookie > accept language
+            // Change the language to 'en-US'
+            assertThat(ctx2.changeLang("en-US")).isTrue();
+            // The messages language 'en-US'
+            assertThat(ctx2.messages().lang().code()).isEqualTo("en-US");
```
Comment (mkurz) [inline framework/src/play-java/src/test/java/play/mvc/HttpTest.java]: @schmitch Please add a test scenario like the last one but instead of using `.changeLang(...)` use `ctx3.setTransientLang(...)`? `setTransientLang`, unlike `changeLang`, doesn't even set a cookie but *only* the variable so we are 100% sure `message()` takes the variable into account as well and not just reads the cookie.

Comment (schmitch) [reply framework/src/play-java/src/test/java/play/mvc/HttpTest.java]: added

Comment (mkurz) [reply framework/src/play-java/src/test/java/play/mvc/HttpTest.java]: Thanks! To confirm I even ran the tests on my machine.

File: framework/src/play/src/main/scala/play/api/i18n/Messages.scala [hunk 4]
```diff
@@ -34,6 +35,16 @@ import scala.util.parsing.input._
  */
 object Messages extends MessagesImplicits {
 
+  /**
+   * Request Attributes for the MessagesApi
+   * Currently all Attributes are only available inside the [[MessagesApi]] methods.
+   */
+  object Attrs {
+
+    val CurrentContextLang: TypedKey[Lang] = TypedKey("CurrentContextLang")
```
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Could we maybe call this `CurrentLang`? It seems to me this is a more general mechanism that could be useful in Scala as well.

Also I'd prefer to omit the empty lines above and below.

Comment (wsargent) [reply framework/src/play/src/main/scala/play/api/i18n/Messages.scala]: Weren't we going to move to `java.lang.Locale` at some point?
